crates/uv/src/commands/project/lock.rs

```
@@ -180,59 +179,129 @@ pub(super) async fn do_lock(
-    )
-    .await?;
+    // Try resolving from the lockfile.
+    let resolution = match lock {
+        Some(lock) => {
+            // Prefill the index with the lockfile.
+            let index = lock.to_index(workspace)?;
+
+            // Create a build dispatch.
+            //
+            // When resolving from the lockfile we can still download/install new distributions,
+            // but we rely on the lockfile for the metadata of any existing distributions. If we
+            // have any outdated metadata we fall back to a clean resolve.
+            let build_dispatch = BuildDispatch::new(
+                &client,
+                cache,
+                interpreter,
+                index_locations,
+                &flat_index,
+                &index,
+                &git,
+                &in_flight,
+                index_strategy,
+                setup_py,
+                config_setting,
+                build_isolation,
+                link_mode,
+                build_options,
+                exclude_newer,
+                concurrency,
+                preview,
+            );
+
+            // TODO(ibraheem): avoid cloning the manifest here
+            pip::operations::resolve(
+                requirements.clone(),
+                constraints.clone(),
+                overrides.clone(),
+                dev.clone(),
+                source_trees.clone(),
+                None,
+                &extras,
+                preferences.clone(),
+                EmptyInstalledPackages,
+                &hasher,
+                &Reinstall::default(),
+                upgrade,
+                None,
+                None,
+                python_requirement.clone(),
+                &client,
+                &flat_index,
+                &index,
+                &build_dispatch,
+                concurrency,
+                options,
+                printer,
+                preview,
+            )
+            .await
+            .ok()
```

Right now this assumes that `Lock::from_resolution_graph` won't fail on the resolution from a lockfile where it would otherwise succeed.. I think that makes sense but now 100% sure.

---

That's plausible I think, but definitely feels pretty difficult to guarantee. There are a lot of error cases.

What happens if the assumption is wrong?

---

If it's wrong we run into error cases because the metadata we construct from the lock file is invalid, and don't fallback to a clean resolve. However, silently falling back would mean potentially hiding bugs in `lock.to_index`, so I think this is fine?